### PR TITLE
[router] Support platform extensions for _layout and routes

### PR DESCRIPTION
**Pull request by marklawlor**
# Why

Revised version #26047
Blocked by #27407 

# How

Rules:

* API routes cannot have platform extensions
* Layouts and routes can have extensions only if a non-extension version exists

For example, you have have `/page.web.tsx` only if `/page.tsx` also exists. This makes platform extensions an additive change and won't break concepts like universal deep linking and typed routes

API routes do not support platform extensions as they are independent of a platform.

# Implementation

`getRoutes` allows multiple files to be assigned to the same route, but all files were given the same specificity.

This PR adds logic to calculate the specificity of a file and to select the most specific file.

Specificity highest-to-lowest

`2`: Platform extension matches platform
`1`: `.native` platform extension
`0`: No platform extensions
`-1`: Invalid route 

If no platform is provided to `getRoutes` the current behaviour is retained and files with platform extensions will be ignored.

# `getRoutes` checklist

These are the locations we use `getRoutes` and whether they should use platform extensions or not.

**[renderStaticContent](https://github.com/expo/expo/blob/269625f5b70a95656212ad14ca7f3e82b45b9f61/packages/expo-router/src/static/renderStaticContent.tsx#L19)**

- Used in `getManifest()` ❌
- Used in `getBuildTimeServerManifestAsync` ❌ - I'm not 100% sure on this one. Need to understand how to test this correctly

**[routes-manifest](
https://github.com/expo/expo/blob/269625f5b70a95656212ad14ca7f3e82b45b9f61/packages/expo-router/src/routes-manifest.ts#L4)**

- Used in `createRoutesManifest` ❌

**[router-store](https://github.com/expo/expo/blob/269625f5b70a95656212ad14ca7f3e82b45b9f61/packages/expo-router/src/global-state/router-store.tsx#L27)**

- Used in `RouterStore` ✅

**[typed-routes](https://github.com/expo/expo/blob/269625f5b70a95656212ad14ca7f3e82b45b9f61/packages/expo-router/src/typed-routes/generate.ts#L5)**

- Used in `getTypedRoutesDeclarationFile` ❌

There is also an `getExactRoutes` function, but this is only used for testing purposes.

** 

# Test Plan

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-router/src/__tests__/platform-routes.test.web.ts [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+  it('pages', () => {
+    expect(() => {
+      getRoutes(
+        inMemoryContext({
+          './folder/page.web.tsx': () => null,
+        }),
+        { internal_stripLoadRoute: true, platform: Platform.OS, skipGenerated: true }
+      );
+    }).toThrowErrorMatchingInlineSnapshot(
+      `"The file ./folder/page.web.tsx does not have a fallback sibling file without a platform extension."`
```
Comment (EvanBacon) [inline packages/expo-router/src/__tests__/platform-routes.test.web.ts:166]:
```suggestion
      `"Platform-specific route "./folder/page.web.tsx" is missing the required fallback route "./folder/page.tsx"."`
```

**Comment by marklawlor** [comment]
Lock `getManifest` / `getBuildTimeServerManifestAsync` / `createRoutesManifest` to platform `web`

File: packages/expo-router/src/__tests__/getRoutes.test.ios.ts [hunk 2]
```diff
@@ -769,6 +769,25 @@ describe('api routes', () => {
...
+      }),
+      {
+        internal_stripLoadRoute: true,
+        platform: 'web',
+        skipGenerated: true,
+        preserveApiRoutes: true,
+      }
+    );
+  }).toThrowErrorMatchingInlineSnapshot(
+    `"Api routes cannot have platform extensions. Please remove '.web' from './folder/two+api.web.tsx'"`
```
Comment (EvanBacon) [inline packages/expo-router/src/__tests__/getRoutes.test.ios.ts:787]:
```suggestion
    `"API routes do not support platform extensions. Remove '.web' from './folder/two+api.web.tsx'"`
```

**Review by EvanBacon — APPROVED** [review]

**Comment by marklawlor** [comment]
Add flag to disable.
Merge into next SDK

**Pull request merged**